### ISSUE TITLE
[rvdm] Failure to halt hart when debugging over JTAG

### ISSUE THREAD

**Issue by jwnrt** [lowRISC/opentitan#19339]
## Summary

Sometimes when trying to debug Tock, OpenOCD will fail to connect and report that it cannot halt the hart.

Here is the console log from OpenOCD with an error message (seems to match [this issue](https://github.com/lowRISC/ibex-demo-system/issues/48)):

```console
[jw opentitan]$ ./bazelisk.sh run //third_party/openocd -- \                                         
    -f "bazel-opentitan/$(./bazelisk.sh outquery //third_party/openocd:jtag_adapter_cfg)" \
    -c "adapter speed 500; transport select jtag; reset_config trst_only" \
    -f util/openocd/target/lowrisc-earlgrey.cfg
[...]
Open On-Chip Debugger 0.12.0
Licensed under GNU GPL v2
For bug reports, read
	http://openocd.org/doc/doxygen/bugs.html
trst_only separate trst_push_pull

Info : Hardware thread awareness created
force hard breakpoints
Info : Listening on port 6666 for tcl connections
Info : Listening on port 4444 for telnet connections
Info : clock speed 500 kHz
Info : JTAG tap: riscv.tap tap/device found: 0x10001cdf (mfg: 0x66f (lowRISC), part: 0x0001, ver: 0x1)
Info : datacount=2 progbufsize=8
Info : Examined RISC-V core; found 1 harts
Info :  hart 0: XLEN=32, misa=0x40901104
Info : starting gdb server for riscv.tap.0 on 3333
Info : Listening on port 3333 for gdb connections
Info : accepting 'gdb' connection on tcp/3333
Info : [riscv.tap.0] Found 4 triggers
Error: unable to halt hart 0
Error:   dmcontrol=0x80000001
Error:   dmstatus =0x00030c82
Info : dropped 'gdb' connection
```
## Context

We can reliably connect while ibex is spinning in a `wfi` loop, but other times it will often fail. It's difficult to tell where the code is when it fails since we can't debug it.

We have also observed that setting a breakpoint and continuing to it (or using the `next` command in GDB) will sometimes take a very long time. Interrupting GDB while it's doing this triggers the "unable to halt hart 0" problem again, and then the debugger will not be able to connect again until reset.

Note that ibex is able to run through the code fine while breakpoints are not set. I'm using [this OpenOCD config](https://github.com/lowRISC/opentitan/blob/master/util/openocd/target/lowrisc-earlgrey.cfg#L41) which forces hardware breakpoints.

## Reproducing

I can reliably hit the problem when running the latest Tock on the CW310. 

I will try stepping through until GDB hits a long delay to narrow down the problem. Maybe then we can have a small software reproduction for this, but I don't have one yet.

**Comment by msfschaffner** [comment]
CC @a-will @andreaskurth @GregAC

**Comment by msfschaffner** [comment]
@jwnrt can you check whether this problem still persists with the current setup? It would be great to root cause this and see whether it is a HW issue that we need to fix for PROD, or whether it is "just" a setup issue (similar to the reset sequencing problem we had experienced last year). Thanks!

**Comment by a-will** [comment]
With the usual setup, Tock is running with the watchdog timer enabled (and _locked_) with the "bite" at 1 second, isn't it? That's totally incompatible with debugging, since aon_timer doesn't stop counting when Ibex is halted.

**Comment by jwnrt** [comment]
Yes, I can reproduce with the Tock that's run with `bazel test //sw/device/silicon_owner/tock/tests/basic:basic_test`. If it's caused by Tock hitting the watchdog then I guess that explains it. If the watchdog can be turned off in Tock easily then it might be worth confirming that's the issue, but I would have to wait until next week to check that myself.

My reproduction steps are here if you want to try yourself:

```sh
sudo apt install openocd
./bazelisk.sh test //sw/device/silicon_owner/tock/tests/basic:basic_test
./bazelisk.sh run //sw/host/opentitantool -- --interface cw310 gpio write TAP_STRAP0 false && bazel run //sw/host/opentitantool -- --interface cw310 gpio write TAP_STRAP1 true
./bazelisk.sh run //third_party/openocd -- \
    -f /usr/share/openocd/scripts/interface/ftdi/olimex-arm-usb-tiny-h.cfg \
    -c "adapter speed 500; transport select jtag; reset_config trst_only" \
    -f util/openocd/target/lowrisc-earlgrey.cfg
```

**Comment by msfschaffner** [comment]
Good point @a-will!
Confirming that'd be great @jwnrt.

**Comment by jwnrt** [comment]
When trying to patch the latest Tock, we found another problem that prevented the debugger connecting. @pamaury worked out that Tock was configuring the pinmux to connect the TAP straps as GPIO pins which breaks trying to use them in RMA mode where they're continuously sampled. We submitted [this PR](https://github.com/tock/tock/pull/3809) to fix that.

My attempts to disable the watchdog have so far not resolved this problem, but I'll keep trying.
